chore(dns): remove `DNS_SERVER`, `DNS_KEEP_NAMESERVER` and replace `DNS_ADDRESS` with `DNS_UPSTREAM_PLAIN_ADDRESSES`

qdm12 opened the pull request:
# Description

Run this with image tag `:pr-2988`. No setting change needed.

Following #2970 plan with a few adjustments:

- Remove `DNS_SERVER` (aka DOT) option: the DNS server forwarder part is now always enabled (see below why)
- Remove `DNS_KEEP_NAMESERVER`: the container will always use the built-in DNS server forwarder, because it can handle now local names with local resolvers (see #2970), it can use the `plain` upstream type (see https://github.com/qdm12/gluetun/commit/5ed6e8292278b54bb5081de0e8ccd0d63a275b3c) AND you can use `DNS_UPSTREAM_PLAIN_ADDRESSES` (see below)
- Replace `DNS_ADDRESS` with `DNS_UPSTREAM_PLAIN_ADDRESSES`:
  - New CSV format with port, for example `ip1:port1,ip2:port2`
  - requires `DNS_UPSTREAM_TYPE=plain` to be set to use `DNS_UPSTREAM_PLAIN_ADDRESSES` (unless using retro `DNS_ADDRESS`)
  - retrocompatibility with `DNS_ADDRESS`. If set, force upstream type to plain and empty user-picked providers. 127.0.0.1 is now ignored since it's always set to this value internally.
  - Warning log on using private upstream resolvers updated
- Warning log if using a private IP address for the plain DNS server which is not in your local subnets
All in all, this greatly simplifies code and available options (less options for the same features is a win). It also allows you to specify multiple plain DNS resolvers on ports other than 53 if needed.

# Issue


Indirectly fix #2840 


# Assertions

* [x] I am aware that we do not accept manual changes to the servers.json file <!-- If this is your goal, please consult https://github.com/qdm12/gluetun-wiki/blob/main/setup/servers.md#update-using-the-command-line -->
* [x] I am aware that any changes to settings should be reflected in the [wiki](https://github.com/qdm12/gluetun-wiki/)

peterfour commented:
Adding to that PR, I've been pulling my hair for a while now with #2970. Had the container set up as a sidecar but it just wouldn't work but somehow was fine as a normal container.
However tried with that pr and it seems to work perfectly fine as a sidecar container.

But I'm still having an issue where from within the pod with the vpn, the other container can't seem to resolve service name from other pods

peterfour commented:
@mglazenborg well I'll be damned, I just tried with `:latest` as well and now everything is working. :tada: 

sidecar vpn in the pod is all good and other container are behind it
and I can use the shorthand internal service name

EDIT: Spoke too soon actually most of it is working but looks like I'm missing dns resolution in the container in the pod somehow

hitem commented:
`DNS_UPSTREAM_TYPE: 'plain' `
`DNS_UPSTREAM_PLAIN_ADDRESSES: '192.168.1.10' `
(also tried specific port :53, and i tried using multiple ip:port as i have multiple dns servers in my network).
Unfortunately its not working for me. it does not use my local dns in my network (192.168.1.10 (ie local address). I see no request coming through. I have not configured or touched anything within the container such as the resolver or anything as such adjustments was not mentioned.
in the startup terminal log i see:
`├── DNS settings:`
`|   ├── Keep existing nameserver(s): no`
`|   ├── DNS server address to use: 127.0.0.1`
`|   └── DNS forwarder server enabled: no`
`2025-YY-XXTXX:XX:XX+XX:00 INFO [dns] using plaintext DNS at address 1.1.1.1`
`2025-YY-XXTXX:XX:XX+XX:00 INFO [vpn] You are running on the bleeding edge of latest!`

going back to setting the "soon to be legacy"
`DNS_KEEP_NAMESERVER: 'on'`
`DNS_SERVER:  '192.168.1.10'`

I once again see my DNS requests coming through my DNS server on 192.168.1.10.

If it helps at all, i do use mullvad config and i do use DOH through my own dns servers.

mglazenborg commented:
Seems to be working as expected now. `/etc/resolv.conf` has 127.0.0.1 set and I'm able to resolve both internal kubernetes records as well as external records (e.g. google.com).

Only thing that isn't working is search domains, but imo that's not a dealbreaker.

qdm12 commented:
@mglazenborg 

> Only thing that isn't working is search domains, but imo that's not a dealbreaker.

Interesting. I did some initial work on search domains and then deleted it 😄 Because I thought it would be the OS/other programs doing the queries using the search option from /etc/resolv.conf, not something the dns resolver would do. I guess I might had been wrong! Thoughts on that?

---

@hitem 

> using plaintext DNS at address 1.1.1.1

Is just whilst waiting for the DNS server to be setup. I'll add another log line precising now the dns server is in use, but you can also check with `docker exec gluetun cat /etc/resolv.conf`

> And i do not see any requests coming through my DNS server on the network (192.168.1.100)-

I'll have a look and add some debug logs for you to check (today/tomorrow)

---

@denniszl thanks for confirming!

hitem commented:
> @mglazenborg
> 
> > Only thing that isn't working is search domains, but imo that's not a dealbreaker.
> 
> Interesting. I did some initial work on search domains and then deleted it 😄 Because I thought it would be the OS/other programs doing the queries using the search option from /etc/resolv.conf, not something the dns resolver would do. I guess I might had been wrong! Thoughts on that?
> 
> @hitem
> 
> > using plaintext DNS at address 1.1.1.1
> 
> Is just whilst waiting for the DNS server to be setup. I'll add another log line precising now the dns server is in use, but you can also check with `docker exec gluetun cat /etc/resolv.conf`
> 
> > And i do not see any requests coming through my DNS server on the network (192.168.1.100)-
> 
> I'll have a look and add some debug logs for you to check (today/tomorrow)
> 
> @denniszl thanks for confirming!

Cheers man. Today i did some more testing (with my limited dns and docker desktop knowledge).
When i swapped my ZeroOmega proxy in browser to use the systemproxy instead (same as windows) it all worked and i see requests coming through, so i started testing the pr2988 again without any success, but maybe something below helps in the troubleshooting?

I ran this command inside my container using the pr2988 and DNS_UPSTREAM_PLAIN_ADDRESSES etc:

```
/ # env | egrep -i 'DNS|GLUETUN|DOT|DOH'
DNS_BLOCK_IPS=
PUBLICIP_FILE=/tmp/gluetun/ip
DNS_BLOCK_IP_PREFIXES=
HTTP_CONTROL_SERVER_AUTH_CONFIG_FILEPATH=/gluetun/auth/config.toml
DNS_UPSTREAM_IPV6=off
STORAGE_FILEPATH=/gluetun/servers.json
DNS_UPSTREAM_RESOLVER_TYPE=DoT
VPN_PORT_FORWARDING_STATUS_FILE=/tmp/gluetun/forwarded_port
DNS_UPDATE_PERIOD=24h
DNS_UPSTREAM_PLAIN_ADDRESSES=192.168.1.100:53
DNS_UNBLOCK_HOSTNAMES=
DNS_UPSTREAM_RESOLVERS=cloudflare
DNS_CACHING=on
DNS_UPSTREAM_TYPE=plain
DNS_REBINDING_PROTECTION_EXEMPT_HOSTNAMES=
```

Here it seems upstream resolver comes back as cloudflare and dot, even tho i dont set those in my glutun config, so i tested just to force it away (it seems these are set even when using latest and DNS_ADRESS etc btw).
so i tried setting:
      DNS_UPSTREAM_RESOLVER_TYPE: "plain" 
      DNS_UPSTREAM_RESOLVERS: "" 

in the config of my container, but then i have these errors:

```
2026-02-01T09:33:28+01:00 INFO [vpn] There is a new release v3.41.0 (v3.41.0) created 38 days ago 
2026-02-01T09:33:37+01:00 INFO [dns] downloading hostnames and IP block lists 
2026-02-01T09:33:38+01:00 INFO [dns] falling back on plaintext DNS at address 192.168.1.100
2026-02-01T09:33:38+01:00 WARN [dns] building server settings: creating plain DNS dialer: validating settings: upstream resolver 192.168.1.100:53: DoT server IPv4 addresses not set 
2026-02-01T09:33:38+01:00 INFO [dns] attempting restart in 20s 
```

over and over (the attempt restart timer increase each attempt). i can still browse the internet etc and no dns comes through to my dns server. I also tried settings DOT: Off or not having it present. same issue.

```
❯ docker exec -it egress-mycontainerherelmao sh -lc 'env | egrep -i "DNS_UPSTREAM|DNS_ADDRESS|DNS_KEEP|DOT|DOH"'
DNS_UPSTREAM_IPV6=off
DNS_UPSTREAM_RESOLVER_TYPE=plain
DNS_UPSTREAM_PLAIN_ADDRESSES=192.168.1.100:53
DNS_UPSTREAM_RESOLVERS=
DNS_UPSTREAM_TYPE=plain
```
     environment:
      DNS_UPSTREAM_PLAIN_ADDRESSES: 192.168.1.100:53
      DNS_UPSTREAM_RESOLVER_TYPE: plain
      DNS_UPSTREAM_RESOLVERS: ""
      DNS_UPSTREAM_TYPE: plain
      HTTP_CONTROL_SERVER_ADDRESS: 0.0.0.0:8000
      HTTP_CONTROL_SERVER_LOG: "off"
      HTTPPROXY: "on"
      HTTPPROXY_LOG: "off"
      HTTPPROXY_PORT: ""
      HTTPPROXY_STEALTH: "off"
      HTTPPROXY_USER: ""

Going back to the latest distro and just using
     DNS_ADDRESS=192.168.1.100
     DNS_KEEP_NAMESERVER=on

Everything works again, i see dns coming through my dns server, the output:

```
/ # env | egrep -i 'DNS_UPSTREAM|DNS_ADDRESS|DNS_KEEP|DOT|DOH'
DNS_UPSTREAM_IPV6=off
DNS_UPSTREAM_RESOLVER_TYPE=DoT
DNS_ADDRESS=192.168.1.100
DNS_KEEP_NAMESERVER=on
DNS_UPSTREAM_RESOLVERS=cloudflare

```

ynnckvdv commented:
Hi, currently running into a similar issue where I cannot resolve to other Kubernetes services but it can reach the internet. I see this is a work in progress, is there anything (experimental or not) I can try right now? 

Then I can connect my Kubernetes services to each other? I've read I can turn `DNS_KEEP_NAMESERVER` on, but this could result in DNS-leaks. These are my DNS-environment variables:

```
/ # env | egrep -i 'DNS|GLUETUN|DOT|DOH'
DNS_BLOCK_IPS=
PUBLICIP_FILE=/tmp/gluetun/ip
DNS_BLOCK_IP_PREFIXES=
HTTP_CONTROL_SERVER_AUTH_CONFIG_FILEPATH=/gluetun/auth/config.toml
DNS_UPSTREAM_IPV6=off
STORAGE_FILEPATH=/gluetun/servers.json
DNS_SERVER=on
DNS_UPSTREAM_RESOLVER_TYPE=DoT
DNS_ADDRESS=127.0.0.1
DNS_KEEP_NAMESERVER=off
VPN_PORT_FORWARDING_STATUS_FILE=/tmp/gluetun/forwarded_port
DNS_UPDATE_PERIOD=24h
DNS_UNBLOCK_HOSTNAMES=
DNS_UPSTREAM_RESOLVERS=cloudflare
DNS_CACHING=on
DNS_REBINDING_PROTECTION_EXEMPT_HOSTNAMES=
```
/etc/resolv.conf :
```
/ # cat /etc/resolv.conf
nameserver 127.0.0.1
search my-namespace.svc.cluster.local svc.cluster.local cluster.local search.domain
options ndots:5
```

If it's better to open a new issue, let me know, but it looked like it would fit here. Cheers guys!

qdm12 commented:
@hitem 

> so i started testing the pr2988 again without any success

And I did write a pile of shitty code that made absolutely no sense, fixed in 37bc4158fb1351bd49b42e3440b3cd98e09ea503 - that should fix it I think!

> building server settings: creating plain DNS dialer: validating settings: upstream resolver 192.168.1.100:53: DoT server IPv4 addresses not set 

Fixed in https://github.com/qdm12/dns/commit/6dbc4b280dc6b625054490539c3f53fc75c9f9d5 and dns bumped to have this commit in this PR (3f71e65d90500a371e51cd22389d67d868552b9c)

---

@mglazenborg 

> Did some furher checks and this might be specific to Alpine Linux, as it seems to work with other distros (Ubuntu/SUSE).

So overall, it's just Alpine Linux not searching correctly on other namespaces right, so all good on "our" side right? Feel free to open an issue or comment here in case you find a workaround for alpine though, since that may be problematic for k8s usage I suppose right?

---

@ynnckvdv the `latest` image should be able to resolve local names now; if not what debug logs do you get? Make sure you pull the latest image. I think v3.41 got this feature as well if I remember correctly. Or you can try that image `:pr-2988` 🤷
